rust/arrow/src/array/transform/mod.rs

```diff
@@ -714,6 +714,48 @@ mod tests {
         assert_eq!(array, expected)
     }
 
+    #[test]
+    fn test_struct_offset() {
+        let strings: ArrayRef = Arc::new(StringArray::from(vec![
+            Some("joe"),
+            None,
+            None,
+            Some("mark"),
+            Some("doe"),
+        ]));
+        let ints: ArrayRef = Arc::new(Int32Array::from(vec![
+            Some(1),
+            Some(2),
+            Some(3),
+            Some(4),
+            Some(5),
+        ]));
+
+        let array =
+            StructArray::try_from(vec![("f1", strings.clone()), ("f2", ints.clone())])
+                .unwrap()
+                .slice(1, 3)
+                .data();
+        let arrays = vec![array.as_ref()];
+        let mut mutable = MutableArrayData::new(arrays, false, 0);
+
+        mutable.extend(0, 1, 3);
+        let data = mutable.freeze();
+        let array = StructArray::from(Arc::new(data));
+
+        // Struct equality doesn't seem to work when using slices?
```

Comment:
This has me suspicious of something not working. Specifically, using equality on the struct or the strings doesn't seem to work as I would expect.

Reply:
Related to #9211 . When we `slice` an array, we increase its `offset`. However, for StructArrays, I think that we also need to increase the offset of the childs (or something like that). :(

Reply:
More specifically with @nevi-me 's analysis [here](https://github.com/apache/arrow/pull/9211#issuecomment-761743850)

Reply:
That makes sense. It seems like it may also affect the string array. When I create the expected string directly instead of slicing it this test passes (see newest commit).

Reply:
I am sorry, I was not very specific. There are two different issues: for string arrays, binary, etc., I think that the issue is solved with #9211, via https://github.com/apache/arrow/pull/9211/files#diff-74d8df3798e724950c2eb5aae1a838fc2f2b9e35d89ace2627e8e7a64584c71fR91

For the Struct, we need another patch.

Reply:
Ah. That makes sense. Thanks for the clarification!

It's unrelated to this PR but I was also wondering about the implementation of extend for structs. For other types, it seems like it is relatively efficient (extending with an entire slice) but for the struct case it seems like it iterates over each row and extends a single row at a time. Is that an actual issue? Would addressing that require the changes to struct you mention?